test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,1167 @@
+    basic_type = dtp->DTP_type_signature.DTP_pool_basic.DTP_basic_type;
+    basic_type_count = dtp->DTP_type_signature.DTP_pool_basic.DTP_basic_type_count;
+    err = MPI_Type_size(basic_type, &basic_type_size);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    size = basic_type_size * basic_type_count;
+    buf = (void *) malloc(size);
+    if (!buf) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+    memset(buf, 0, size);
+
+    /* create contiguous datatype */
+    obj_idx = par->obj_idx;
+    err = MPI_Type_contiguous(basic_type_count, basic_type, &dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    err = MPI_Type_commit(&dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    dtp->DTP_obj_array[obj_idx].DTP_obj_count = 1;
+
+    /* initialize buffer with requested params */
+    err = MPI_Type_get_name(basic_type, basic_type_name, &len);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    DTPI_Type_init_buf(par, basic_type_name, buf);
+
+    dtp->DTP_obj_array[obj_idx].DTP_obj_buf = buf;
+    dtp->DTP_obj_array[obj_idx].private_info = dtpi;
+
+    /* initialize private datatype info data */
+    dtpi->obj_type = DTPI_OBJ_TYPE__CONTIG;
+    dtpi->type_basic_size = basic_type_size;
+    dtpi->type_extent = basic_type_size;
+    dtpi->type_lb = 0;
+    dtpi->type_ub = basic_type_size * basic_type_count;
+
+    /* set type name for debug information */
+    sprintf(type_name, "%s (%s basic_type %li count)", "contig", basic_type_name, basic_type_count);
+    err = MPI_Type_set_name(dtp->DTP_obj_array[obj_idx].DTP_obj_type, (char *) type_name);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+
+  fn_exit:
+    return err;
+
+  fn_fail:
+    if (dtpi) {
+        free(dtpi);
+    }
+    goto fn_exit;
+}
+
+int DTPI_Vector_create(struct params *par, DTP_t dtp)
+{
+    int err = DTP_SUCCESS;
+    int len, obj_idx;
+    int basic_type_size;
+    MPI_Aint type_count, type_stride, type_blklen;
+    MPI_Aint size;
+    MPI_Datatype basic_type;
+    void *buf;
+    DTPI_t *dtpi = NULL;
+    char type_name[TYPE_NAME_MAXLEN] = { 0 };
+    char basic_type_name[BASIC_TYPE_NAME_MAXLEN] = { 0 };
+
+    /* allocate space for private datatype info */
+    dtpi = (DTPI_t *) malloc(sizeof(*dtpi));
+    if (!dtpi) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    /* allocate buffer */
+    basic_type = dtp->DTP_type_signature.DTP_pool_basic.DTP_basic_type;
+    type_count = par->type_count;
+    type_stride = par->type_stride;
+    err = MPI_Type_size(basic_type, &basic_type_size);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    size = type_count * type_stride * basic_type_size;
+    buf = (void *) malloc(size);
```

Comment:
- Same here, please use `MPI_Type_get_extent` instead of manually calculating buffer size.
- Why `size = count * stride  * sizeof(basic_element)` ? It seems incorrect.
- Instead of set `par->type_count` to a local variable, why not use it directly ? This can hint that this is an input parameter. Plus it can reduce count of lines.

Reply:
Ok I will update it